### PR TITLE
functionalization fix for inplace comparison ops

### PR DESCRIPTION
**Pull request by bdhirsh**
This is an interesting bug that surfaced from trying to integrate with LTC.

If you look at the `torch.ge` function, its dtype promotion logic diverges between the functional and inplace variant:

(1) `torch.ge(a, b)` always returned a `bool` tensor
(2) `a.ge_(b)` won't change the dtype of `a`. So the returned tensor is whatever `a`'s dtype is

This means that if a user calls `a.ge_(b)` and we want to functionalize it into an `at::ge(a, b)` call, then the metadata on the inner tensor inside of `FunctionalTensorWrapper` will be wrong! When we eventually pop out of functionalization (and return the inner tensor back to the user), it will have the wrong dtype.

That actually means that the "correct" transformation for `ge_` would be:

**Before**
```
$1 = torch._ops.aten.ge_.Scalar($0, 0)
```
**After**
Manually perform a dtype cast afterwards if metadata from the functional op call is "wrong"
```
$1 = torch._ops.aten.ge.Scalar($0, 0)
$2 = torch._ops.aten._to_copy.default($1, dtype=6, layout=0)""")
```

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75527
* #76358
* #77714
* #77713
* #77132
* #77716
* #77130
* #77129
* #77128
* __->__ #77125
* #77126

### CODE REVIEW

File: aten/src/ATen/FunctionalTensorWrapper.cpp [hunk 1]
```diff
@@ -183,6 +183,10 @@ void FunctionalTensorWrapper::replace_(const Tensor& other) {
   // out= ops are allowed to resize the output tensors, mutating both the data and metadata of the tensor.
   // We need to propagate that metadata mutation to the wrapper (new size).
   set_sizes_and_strides(value_.sizes(), value_.strides());
+  set_storage_offset(value_.storage_offset());
+  if (dtype() != value_.unsafeGetTensorImpl()->dtype() || layout() != value_.unsafeGetTensorImpl()->layout()) {
+    value_ = value_.to(c10::TensorOptions().dtype(dtype()).layout(layout()));
+  }
```
Comment (bdhirsh) [inline aten/src/ATen/FunctionalTensorWrapper.cpp:189]: Right now I do this by making assumptions about which classes of metadata can be mutated by an operator
- size/stride/storage offset can all be modified by an inplace op (e.g. out= ops that resize, or transpose_())
- dtype/layout are never modified by an operator, which lets me assume that the dtype/layout on the wrapper are "correct", and that I can propagate them to the inner tensor if they diverge

Comment (bdhirsh) [reply aten/src/ATen/FunctionalTensorWrapper.cpp:189]: With this change (and some more further in the stack, I was able to get all of the LTC tests to pass, which makes me reasonable convinced that I haven't missed any weird cases.

Now that I think about it though, the absolutely correct thing to do would probably just to run the original operator with meta tensors, and compare the output meta tensor's metadata to the functional op's output.

We can't actually do that though without full meta tensor coverage.

**Comment by ezyang** [comment]
cc @mruberry @ngimel, this is the kind of thing primtorch wants to handle right

File: aten/src/ATen/FunctionalTensorWrapper.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/FunctionalTensorWrapper.cpp:189]: Do you need to do this with memory format too?

**Review by ezyang — APPROVED** [review]

**Comment by mruberry** [comment]
> cc @mruberry @ngimel, this is the kind of thing primtorch wants to handle right

I think it will; we intend to model inplace operations as a safe copy to the out tensor after the operation is performed, which is kinda what functionalization wants to do in the first place? This is how we model out= today

File: aten/src/ATen/FunctionalTensorWrapper.cpp (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/FunctionalTensorWrapper.cpp:189]: Yes doing this with memory format will be good. 
Another thought (that's for all functionalized inplace ops) - if `a` is broadcasted, inplace op will balk, but `copy_to` will happily proceed.

Comment (bdhirsh) [reply aten/src/ATen/FunctionalTensorWrapper.cpp:189]: Is there... a good way to check if two tensors have the same memory format? (what if they're both channels last, but one is contiguous and the other isn't?).

I guess I was thinking something like this, which looks pretty awful:
```
auto other_chan_last = other.unsafeGetTensorImpl()->is_strides_like_channels_last();
auto same_channels_last = is_strides_like_channels_last() == other_chan_last;

auto other_chan_last_3d = other.unsafeGetTensorImpl()->is_strides_like_channels_last_3d();
auto same_channels_last_3d = is_strides_like_channels_last_3d() == other_chan_last_3d

if (!same_channels_last || !same_channels_last_3d) {
    auto mem_format = other_chan_last ? MemoryFormat::ChannelsLast : other_chan_last_3d ? MemoryFormat::ChannelsLast3d : MemoryFormat::Preserve;
     value_ = value_.to(c10::TensorOptions().dtype(dtype()).layout(layout()), mem_format)
}

```

Comment (ezyang) [reply aten/src/ATen/FunctionalTensorWrapper.cpp:189]: Do they have the same sizes? One way is to compare strides

**Pull request closed without merging**